### ISSUE TITLE
ActivityPub Video `attributedTo` field malformed

### ISSUE THREAD

**Issue by koehn** [Chocobozzz/PeerTube#2704]
**What happened?**
When fetching videos, the `attributedTo` looks like this:
```
  "attributedTo": [
    {
      "type": "Person",
      "id": "https://peertube.koehn.com/accounts/bkoehn"
    },
    {
      "type": "Group",
      "id": "https://peertube.koehn.com/video-channels/39ddd8e5-1938-410b-9d4d-6d64d79f82fa"
    }
  ],
```
I believe this is malformed; the objects in attributedTo should be either complete objects, `Link`s, or ids. Honestly I'd be happy with just the author; this is how most ActivityPub applications handle it. 

**What do you expect to happen instead?**
```
  "attributedTo": [
   "https://peertube.koehn.com/accounts/bkoehn",
   "https://peertube.koehn.com/video-channels/39ddd8e5-1938-410b-9d4d-6d64d79f82fa"
  ],
```

**Steps to reproduce:**

1. `curl -H 'Accept: application/ld+json' https://peertube.koehn.com/videos/watch/72fde0c4-933f-4148-aced-921f362c970d`

2.

3.

**Additional information**
* PeerTube version or URL: `chocobozzz/peertube:production-buster`
* Browser name/version: n/a
* NodeJS version: `chocobozzz/peertube:production-buster`

* Link to browser console log if useful:
* Link to server log if useful (journalctl or /var/www/peertube/storage/logs/):

**Comment by tcitworld** [comment]
What are complete objects? An AS object "needs" basically only `type` and `id` properties. That's the JSON-LD part, following what's at the URI behind `id` to have the full object is up to your implementation.
This form allows knowing right away what's the type of actors without an extra fetch. It may not be the best, but it's not malformed.
https://www.w3.org/TR/activitystreams-vocabulary/#dfn-attributedto also shows a similar form.

**Comment by koehn** [comment]
How is a client to know that a server only sends part of the object, and that they should request the information again to get the complete object? There’s no information there indicating that the object is incomplete. 

I think it would be better to send a link object with an additional field; that would allow clients that know about PeerTube to look for that field, and clients that don’t would know that they’d have to pull the `href` to actually get the object.

**Comment by Chocobozzz** [comment]
Hello,

We should handle `attributedTo: string[]` with https://github.com/Chocobozzz/PeerTube/commit/fccda976c26197ae52c9acafc31b3f6668898a74
In the future, we'll be able to add a federation breaking change by correctly formatting this `attributedTo` field

**Comment by Chocobozzz** [comment]
Sorry for the delay. We should correctly format `attributedTo` in v7.3: https://github.com/Chocobozzz/PeerTube/commit/47426d3610d8d31ff98abc84a8b609c156486320

**Comment by Nutomic** [comment]
Unfortunately this change broke federation with Lemmy.

**Comment by Chocobozzz** [comment]
I'm really sorry about that @Nutomic 

Can you fix this in next Lemmy versions? It's difficult for us to re-introduce the old behaviour since it was an incorrect AP definition :/

**Comment by Nutomic** [comment]
Yes I made a fix, but unfortunately it is a relatively complicated change and the current stable Lemmy version 0.19 is in maintenance mode. So backporting the fix is not practical, and we need to wait for the new version 1.0 to deploy the fix which will take 6-12 more months.

Afaik Peertube is the only platform which uses this specific format for `attributedTo`, which means Lemmy needs [extra code to handle this](https://github.com/LemmyNet/lemmy/blob/main/crates/apub/objects/src/protocol/page.rs#L245). Seeing as you're willing to make breaking changes in this area, it would be ideal if you could instead federate in the way which is already used by Lemmy, Piefed, Mbin and Friendica. Namely put the `Group` id in `to` and only the `Person` id in `attributedTo`. With this change Peertube would immediately be compatible with the stable Lemmy version and all other platforms that federate with Lemmy, and I could remove the extra code for Peertube compat.

**Comment by Chocobozzz** [comment]
Thank you, @Nutomic, for your message.

I'll try to find time to implement your solution in the following weeks.

**Comment by Nutomic** [comment]
@Chocobozzz Hi, any update on this?

**Comment by Chocobozzz** [comment]
No sorry, I've been busy these past few weeks. However, this issue remains at the top of my list. I will ping you when I make progress on this matter.

**Comment by Chocobozzz** [comment]
Reverting to fix Lemmy compatibility. We'll move to a raw `attributedTo` string in v9. 

https://github.com/Chocobozzz/PeerTube/commit/91680e89439fee765250c63f12389507f89052ee
